runtime/compiler/control/HookedByTheJit.cpp

```diff
@@ -2495,11 +2550,8 @@ void jitClassesRedefined(J9VMThread * currentThread, UDATA classCount, J9JITRede
             deserializer->invalidateClass(currentThread, classPair->oldClass);
          }
```

Comment:
I can't easily tell. It seems pretty reasonable that we'd invalidate AOT-related cache entries for the old class (which is always `classPair->oldClass`). If the redefinition was not in-place, then the old class is stale anyway. If OTOH it was in-place, the old class is fresh, but if it was known before redefinition for AOT purposes, then it would have been known with the then-current bytecode, which is now stale. So it makes sense to me on a high level. But I'm not sure in detail. One thing that's confusing to me is that `JITServerNoSCCAOTDeserializer::invalidateClass()` seems to invalidate cache entries for each of the methods of the given class. If it's looking at the old class and redefinition was in-place, then I think by the time we reach here it would be invalidating entries for the fresh methods when I would have expected it to want to invalidate the stale ones instead

Reply:
Should we put this aside as something for @mpirvu, perhaps, to follow up on?

Reply:
I think that makes sense. If there's an issue here, it seems to be an independent one because it's unaffected by the problem I'm fixing in this PR and it's also unaffected by the fix

Reply:
OK, that's what I thought too. I guess the only thing then would be to check on the further method invalidation in `JITServerNoSCCAOTDeserializer::invalidateClass()`, which looks to me like it will currently invalidate the new/fresh methods instead of the old/stale ones when the redefinition was in-place

Reply:
You're saying that if a class is redefined in-place, the methods of `oldClass` will be the new methods and not the old methods? I think that could be a problem, because the deserialization process for a method serialization record looks something like:

1. Look up the cached `J9Class *ramClass` that was earlier associated to the defining class record for that method record. At this point we know that `ramClass->romClass` is identical to the ROMClass of the defining class of the method as it was recorded at compile time.
2. Get the `J9Method *` from `ramClass->ramMethods` at the index specified in the method record and associate it to the method record we're deserializing.

We invalidate the methods of cached classes that are unloaded or redefined because we want to preserve the property that every cached `J9Method *` pointer is still valid and has a defining class with a matching ROMClass as in (1).

Reply:
I'm kind of surprised that this hasn't come up in testing, then, because I think we're inappropriately keeping around stale `J9Method *` pointers in the cache.

Though, now that I look at it, the only place that a cached `J9Method *` can be used is here:

https://github.com/eclipse-openj9/openj9/blob/c895f842e484fbcf5bb349ea2223e2ee8e69d322/runtime/compiler/runtime/RelocationRecord.cpp#L4819-L4834

in the course of that `romMethodFromOffsetInSharedCache` call; that's the only place that method is ever called. But if you look before it, we must already have a `definingClassID` to get to this point in relocation! If I understand what the SVM is doing correctly, that implies that the defining class must have been mentioned in an earlier relocation record, which means that during deserialization we must have ensured that the defining class record was still valid.

So, the fact that we invalidated the `oldClass` is enough to ensure that we never try to access invalid cached methods; we don't support re-caching entries in the `JITServerNoSCCAOTDeserializer`, so invalidated entries remain invalid. Either that, or it must be extremely unlikely that a cached `J9Class *` is redefined in-place and subsequently used in applying a `ValidateMethodFromClassAndSig` record.

Reply:
Will the `newClass` have all the old methods if a class is redefined in place? If so, we can pass its `ramMethods` in to `invalidateClass()` as well. Or, I see this `staleMethod`/`freshMethod` handling in a loop just above the existing `invalidateClass`. If those `staleMethod`s are the ones that are being invalidated, then I can add a `deserializer->invalidateMethod()` to take care of them there.

Reply:
Yes, `newClass` should have the set of methods that `oldClass` doesn't have. For an in-place redefinition, `oldClass` has the new methods and `newClass` has the old ones. For a not-in-place redefinition, `oldClass` has the old methods and `newClass` has the new ones. So if we invalidate cache entries for both sets of methods, we'll be sure to get the old methods either way